core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -1764,11 +1764,14 @@ private[zio] trait ZIOFunctions extends Serializable {
     n: Long
   )(as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] =
     (for {
-      semaphore <- Semaphore.make(n)
-      bs <- foreachPar[R, E, A, B](as) { a =>
-             semaphore.withPermit(fn(a))
-           }
-    } yield bs).refailWithTrace
+      q     <- Queue.bounded[(Promise[E, B], A)](n.toInt)
+      pairs <- ZIO.foreach(as)(a => Promise.make[E, B].map(p => (p, a)))
+      _     <- ZIO.foreach(pairs)(pair => q.offer(pair)).fork
+      _ <- ZIO.collectAll(List.fill(n.toInt)(q.take.flatMap {
+            case (p, a) => fn(a).foldCauseM(c => ZIO.foreach(pairs)(_._1.halt(c)), b => p.succeed(b))
+          }.forever.fork))
+      res <- ZIO.foreach(pairs)(_._1.await).ensuring(q.shutdown)
+    } yield res).refailWithTrace
```

Comment:
There is a fiber-leak potential here: the for comprehension can be interrupted after the queue is created, but before `ensuring(q.shutdown)`. 

To solve this problem, you can use `bracket_`, e.g.:

```scala
Queue.bounded(n.toInt).bracket(_.shutdown) { q =>
  ...
}
```

Reply:
I fixed this.